File: 05/seeds.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

#define BUFFER_SIZE     256
#define MAX_RANGES      200  /* there are ~220 lines total in input, should be enough */
#define MAX_SEEDS       100

struct map_range
{
    long offset;
    long src;
    long limit;
};

struct seed_range
{
    long start;
    long limit;
};

struct seed_range seeds[MAX_SEEDS];
struct map_range seed_to_soil[MAX_RANGES];
struct map_range soil_to_fertilizer[MAX_RANGES];
struct map_range fertilizer_to_water[MAX_RANGES];
struct map_range water_to_light[MAX_RANGES];
struct map_range light_to_temperature[MAX_RANGES];
struct map_range temperature_to_humidity[MAX_RANGES];
struct map_range humidity_to_location[MAX_RANGES];


struct map_range seed_to_soil_cached;
struct map_range soil_to_fertilizer_cached;
struct map_range fertilizer_to_water_cached;
struct map_range water_to_light_cached;
struct map_range light_to_temperature_cached;
struct map_range temperature_to_humidity_cached;
struct map_range humidity_to_location_cached;

static int parse_seeds(char *seeds);
static void read_map(FILE *input, struct map_range *range);
static long lookup(long source, struct map_range *range, struct map_range *cached);

int main(void)
{
    char line[BUFFER_SIZE];
    FILE *input = fopen("input.txt", "r");

    int num_seed_ranges = 0;

    for (;;)
    {
        if (fgets(line, BUFFER_SIZE, input) == NULL) break;

        if (strstr(line, "seeds: ") == line)
            num_seed_ranges = parse_seeds(line + 7);
        else if (strstr(line, "seed-to-soil map:") == line)
            read_map(input, seed_to_soil);
        else if (strstr(line, "soil-to-fertilizer map:") == line)
            read_map(input, soil_to_fertilizer);
        else if (strstr(line, "fertilizer-to-water map:") == line)
            read_map(input, fertilizer_to_water);
        else if (strstr(line, "water-to-light map:") == line)
            read_map(input, water_to_light);
        else if (strstr(line, "light-to-temperature map:") == line)
            read_map(input, light_to_temperature);
        else if (strstr(line, "temperature-to-humidity map:") == line)
            read_map(input, temperature_to_humidity);
        else if (strstr(line, "humidity-to-location map:") == line)
            read_map(input, humidity_to_location);
    }

    fclose(input);

    long total = 0;
    for (int sr = 0; sr < num_seed_ranges; ++sr)
    {
        total += seeds[sr].limit - seeds[sr].start;
    }

    seed_to_soil_cached.limit = 0;
    soil_to_fertilizer_cached.limit = 0;
    fertilizer_to_water_cached.limit = 0;
    water_to_light_cached.limit = 0;
    light_to_temperature_cached.limit = 0;
    temperature_to_humidity_cached.limit = 0;
    humidity_to_location_cached.limit = 0;

    printf("Total: %ld\n", total);

    long min_location = __LONG_MAX__;

    total /= 100;
    long progress = 0;
    long old_percent = 0;

    for (int sr = 0; sr < num_seed_ranges; ++sr)
    {
        for (long seed = seeds[sr].start; seed < seeds[sr].limit; ++seed)
        {
            ++progress;
            long percent = progress / total;
            if (percent != old_percent)
            {
                printf("%d%%\n", percent);
                old_percent = percent;
            }

            long soil = lookup(seed, seed_to_soil, &seed_to_soil_cached);
            long fertilizer = lookup(soil, soil_to_fertilizer, &soil_to_fertilizer_cached);
            long water = lookup(fertilizer, fertilizer_to_water, &fertilizer_to_water_cached);
            long light = lookup(water, water_to_light, &water_to_light_cached);
            long temperature = lookup(light, light_to_temperature, &light_to_temperature_cached);
            long humidity = lookup(temperature, temperature_to_humidity, &temperature_to_humidity_cached);
            long location = lookup(humidity, humidity_to_location, &humidity_to_location_cached);

            if (location < min_location) min_location = location;
        }
    }

    // result p1: 261668924
    // result p2: 24261545

    printf("Result: %ld\n", min_location);

    return 0;
}

static int parse_seeds(char *seed_str)
{
    char *savep;
    char *seed_token = strtok_r(seed_str, " ", &savep);

    int count = 0;
    while (seed_token)
    {
        seeds[count].start = atol(seed_token);
        seed_token = strtok_r(NULL, " ", &savep);
        seeds[count].limit = seeds[count].start + atol(seed_token);

        seed_token = strtok_r(NULL, " ", &savep);
        ++count;
    }

    printf("S: %d\n", count);

    return count;
}

static void read_map(FILE *input, struct map_range *range)
{
    int count = 0;
    long length;
    long dest;

    for (;;++count) {
        int nread = fscanf(input, "%ld %ld %ld", &dest, &range[count].src, &length);
        if (nread != 3) break;
        range[count].limit = range[count].src + length;
        range[count].offset = dest - range[count].src;

        //printf("[%ld:%ld) -> %ld\n", range[count].src, range[count].limit, range[count].dest);
    }
    range[count].limit = 0;
}

static long lookup(long source, struct map_range *range, struct map_range *cached)
{
    // try cached first
    if (source >= cached->src && source < cached->limit)
        return source + cached->offset;

    // find new range
    for (int i = 0; range[i].limit != 0; ++i)
    {
        if (source >= range[i].src && source < range[i].limit)
        {
            memcpy(cached, range + i, sizeof(struct map_range));
            return source + range[i].offset;
        }
    }

    // not found, prepare new pseudo-cache - gap where *source* belongs:
    //     find largest limit which is smaller (or equal) than *source* - use as src
    //     find smallest src, that is larger than *source* - use as limit
    //     the offset remains 0
    cached->offset = 0;
    cached->src = 0;
    cached->limit = __LONG_MAX__;

    for (int i = 0; range[i].limit; ++i)
    {
        if (range[i].limit <= source && cached->src < range[i].limit)
            cached->src = range[i].limit;
        if (range[i].src > source && cached->limit > range[i].src)
            cached->limit = range[i].src;
    }

    return source;
}

File: 05/src/seeds.cs
using System.IO;
using System.Collections.Generic;

public struct MapRange
{
    public long Dest;
    public long Src;
    public long Size;
    public long Limit => Src + Size;
    public long DestLimit => Dest + Size;
    public long Offset => Dest - Src;

}

public struct SeedRange
{
    public long Start;
    public long Size;
    public long Limit => Start + Size;
}

class Program
{
    static void Main()
    {
        using var input = new StreamReader("input.txt");

        var lookup = new SeedLookup(input);


        var min_location = lookup.SearchSmallestLocation();

        // sample result: 35
        // real result: 261668924
        Console.WriteLine($"Result: {min_location}");
    }
}

public class SeedLookup
{
    public readonly List<SeedRange> seeds;
    List<MapRange> seed_to_soil;
    List<MapRange> soil_to_fertilizer;
    List<MapRange> fertilizer_to_water;
    List<MapRange> water_to_light;
    List<MapRange> light_to_temperature;
    List<MapRange> temperature_to_humidity;
    List<MapRange> humidity_to_location;

    public SeedLookup(TextReader input)
    {
        for (;;)
        {
            string? line = input.ReadLine();
            if (line == null) break;

            if (line.StartsWith("seeds: "))
                seeds = ParseSeeds(line.Substring(7));
            else if (line.StartsWith("seed-to-soil map:"))
                seed_to_soil = ReadMap(input);
            else if (line.StartsWith("soil-to-fertilizer map:"))
                soil_to_fertilizer = ReadMap(input);
            else if (line.StartsWith("fertilizer-to-water map:"))
                fertilizer_to_water = ReadMap(input);
            else if (line.StartsWith("water-to-light map:"))
                water_to_light = ReadMap(input);
            else if (line.StartsWith("light-to-temperature map:"))
                light_to_temperature = ReadMap(input);
            else if (line.StartsWith("temperature-to-humidity map:"))
                temperature_to_humidity = ReadMap(input);
            else if (line.StartsWith("humidity-to-location map:"))
                humidity_to_location = ReadMap(input);
        }

        /* to make compiler happy */
        if (seeds == null) seeds = new();
        if (seed_to_soil == null) seed_to_soil = new();
        if (soil_to_fertilizer == null) soil_to_fertilizer = new();
        if (fertilizer_to_water == null) fertilizer_to_water = new();
        if (water_to_light == null) water_to_light = new();
        if (light_to_temperature == null) light_to_temperature = new();
        if (temperature_to_humidity == null) temperature_to_humidity = new();
        if (humidity_to_location == null) humidity_to_location = new();
    }

    public long SearchSmallestLocation()
    {
        long min_location = long.MaxValue;

        foreach (var seed in seeds)
        {
            long soil = Lookup(seed.Start, seed_to_soil);
            long fertilizer = Lookup(soil, soil_to_fertilizer);
            long water = Lookup(fertilizer, fertilizer_to_water);
            long light = Lookup(water, water_to_light);
            long temperature = Lookup(light, light_to_temperature);
            long humidity = Lookup(temperature, temperature_to_humidity);
            long location = Lookup(humidity, humidity_to_location);

            if (location < min_location) min_location = location;
        }

        return min_location;
    }

    static List<SeedRange> ParseSeeds(string seed_str)
    {
        var ranges = seed_str
            .Split(" ")
            .Select(long.Parse)
            .ToArray();

        List<SeedRange> seeds = new List<SeedRange>();
        for (int i = 0; i < ranges.Length; i += 2)
        {
            seeds.Add(new SeedRange {
                Start = ranges[i],
                Size = ranges[i + 1]
            });
        }

        return seeds;
    }

    static List<MapRange> ReadMap(TextReader input)
    {
        List<MapRange> range = new List<MapRange>();

        int count = 0;

        for (;;++count) {
            string? line = input.ReadLine();
            if (string.IsNullOrWhiteSpace(line)) break;
            var values = line.Split(" ").Select(long.Parse).ToList();

            range.Add(new MapRange
            {
                Dest = values[0],
                Src = values[1],
                Size = values[2],
            });
        }

        return FillGaps(range);
    }

    static List<MapRange> FillGaps(IEnumerable<MapRange> range)
    {
        List<MapRange> target = new List<MapRange>();
        long prev_end = 0;
        foreach (var r in range.OrderBy(x => x.Src))
        {
            if (r.Src > prev_end)
                target.Add(new MapRange {
                    Src = prev_end,
                    Dest = prev_end,
                    Size = r.Src - prev_end,
                });
            else if (r.Src < prev_end)
                throw new InvalidOperationException("Range overlap");
            target.Add(r);
            prev_end = r.Limit;
        }

        target.Add(new MapRange {
            Src = prev_end,
            Dest = prev_end,
            Size = long.MaxValue - prev_end,
        });

        return target;
    }

    static long Lookup(long source, List<MapRange> range)
    {
        foreach (var item in range)
        {
            if (source >= item.Src && source < item.Limit)
                return source + item.Offset;
        }

        return source;
    }
}



